File: 2_SEMESTRE/codeJS/3343.js
var input = require('fs').readFileSync('stdin', 'utf-8');
var lines = input.split('\n');
const columns = [];

lines.forEach(l => {
    columns.push(l.split(' '));
})

const heigthWall = Number(columns[0][1]);

const heigthsTitans = [
    {'P': Number(columns[2][0])},
    {'M': Number(columns[2][1])},
    {'G': Number(columns[2][2])},
];

const attacksQueue = columns[1][0].split('');

class Titan {
    constructor(type, heigth) {
        this.type = type,
        this.heigth = heigth,
        this.isAlive = true
    }
    attack(wall) {
        if (this.heigth <= wall.heigth) {
            this.isAlive = false
            wall.lose_heigth(this.heigth)
        }
    }
};

class Wall {
    constructor(heigth) {
        this.heigth = heigth
    }
    lose_heigth(damage) {
        this.heigth = this.heigth - damage
    }
};

class Titans {
    constructor() {
        this.alives = []
    }
    make(type, heigth) {
        const titan = new Titan(type, heigth)
        this.add(titan)
    }
    add(titan) {
        this.alives.push(titan)
    }
};


class Walls {
    constructor() {
        this.built = []
    }
    build() {
        const wall = new Wall(heigthWall)
        this.add(wall)
    }
    add(wall) {
        this.built.push(wall)
    }
};

const walls = new Walls();

for (let i = 0; i < 1; i++) {
    walls.build()
};

const titans = new Titans();

const createAttackTitan = (attacks) => {
    attacks.forEach((attack) => {
        let heigth = heigthsTitans.filter(h => h[`${attack}`])
        titans.make(attack, heigth[0][`${attack}`])
    })
};

createAttackTitan(attacksQueue);

titans.alives.forEach(titan => {
    let lastIndex = 0;
    while (titan.isAlive) {
        walls.built.forEach((wall, index) => {
        if (index >= lastIndex) {
            titan.attack(wall)
            lastIndex = index + 1
        }
        })
        if (titan.isAlive) {
            walls.build()
        } 
    }
});

console.log(walls.built.length);

File: 2_SEMESTRE/codeJS/3343.py
lines: list = []

for _ in range(3):
    line = input()
    lines.append(line)

columns: list = [line.split(' ') for line in lines]

titans_sizes: list[dict[str, int]] = [
    {'P': int(columns[2][0])},
    {'M': int(columns[2][1])},
    {'G': int(columns[2][2])},
]

wall_heigth: int = int(columns[0][1])

titan_attack_queue: list[str] = list(columns[1][0])


class Wall:
    def __init__(self, height: int) -> None:
        self.height = height

    def lose_height(self, damage: int) -> None:
        self.height -= damage


class Titan:
    def __init__(self, _type: str, height: int) -> None:
        self.type = _type
        self.height = height
        self.isAlive = True

    def attack(self, wall: Wall) -> None:
        if self.height <= wall.height:
            self.isAlive = False
            wall.lose_height(self.height)


class Titans:
    def __init__(self) -> None:
        self.alives: list = []

    def make(self, _type: str, height: int) -> None:
        titan: Titan = Titan(_type, height)
        self.add(titan)

    def add(self, titan: Titan) -> None:
        self.alives.append(titan)


class Walls:
    def __init__(self):
        self.built: list = []

    def build(self):
        wall: Wall = Wall(wall_heigth)
        self.add(wall)

    def add(self, wall: Wall):
        self.built.append(wall)


walls: Walls = Walls()

for _ in range(1):
    walls.build()

titans: Titans = Titans()


def create_attack_titan(queue: list[str]):
    for member in queue:
        for size in titans_sizes:
            if member in size:
                height = size[f'{member}']
                titans.make(member, height)


create_attack_titan(titan_attack_queue)

for titan in titans.alives:
    last_index = 0
    while titan.isAlive:
        for index, wall in enumerate(walls.built):
            if index >= last_index:
                titan.attack(wall)
                last_index = index + 1
        if titan.isAlive:
            walls.build()


print(len(walls.built))

lines: list[str] = []
for _ in range(10):
    lines.append(int(input()))

lines = [1 if el <= 0 else el for el in lines if el > 0]

for i, v in enumerate(lines):
    print(f"X[{i}] = {v}")
